tests/daemon/mock-clock.c

```diff
@@ -22,13 +40,89 @@ mock_clock_class_init (MockClockClass *klass)
 {
 }
 
+static gboolean
+mock_clock_source_check (GSource *source)
+{
+  MockClockSource *self = (MockClockSource *) source;
+
+  return self->clock->times[self->type_] >= self->next_ready_time;
+}
+
+static gboolean
+mock_clock_source_dispatch (GSource    *source,
+                           GSourceFunc callback,
+                           gpointer    user_data)
+{
+  MockClockSource *self = (MockClockSource *) source;
+
+  g_return_val_if_fail (callback != NULL, G_SOURCE_REMOVE);
+
+  self->next_ready_time += self->interval;
+  return callback (user_data);
+}
+
+static void
+mock_clock_source_finalize (GSource *source)
+{
+  MockClockSource *self = (MockClockSource *) source;
+
+  g_clear_object (&self->clock);
+}
+
+static GSourceFuncs mock_clock_source_funcs = {
+  .check = mock_clock_source_check,
+  .dispatch = mock_clock_source_dispatch,
+  .finalize = mock_clock_source_finalize,
+};
+
+static GSource *
+mock_clock_source_new (MockClock *self,
+                       MockClockType type_,
+                       guint interval)
+{
+  g_return_val_if_fail (MOCK_IS_CLOCK (self), NULL);
+
+  MockClockSource *source =
+      (MockClockSource *) g_source_new (&mock_clock_source_funcs,
+                                        sizeof (MockClockSource));
+  source->clock = g_object_ref (self);
+  source->type_ = type_;
+  source->interval = interval;
+  source->next_ready_time = self->times[type_] + interval;
+
+  return (GSource *) source;
+}
+
+static GSource *
+mock_clock_timeout_source_new_seconds (EmerClock *self,
+                                       guint      interval)
+{
+  g_return_val_if_fail (MOCK_IS_CLOCK (self), NULL);
+
+  return mock_clock_source_new (MOCK_CLOCK (self),
+                                MOCK_CLOCK_TYPE_MONOTONIC,
+                                interval);
+}
+
 static void
 clock_iface_init (EmerClockInterface *iface,
                   gpointer            iface_data)
 {
+  iface->timeout_source_new_seconds = mock_clock_timeout_source_new_seconds;
 }
 
 static void
 mock_clock_init (MockClock *self)
 {
+  self->times[MOCK_CLOCK_TYPE_MONOTONIC] = 0x1;
+}
+
+void
+mock_clock_advance_monotonic (MockClock *self,
+                              gint64     delta_usecs)
+{
+  g_return_if_fail (delta_usecs >= 0);
```

Comment:
If `delta_usecs` is never allowed to be negative, would it be clearer to make it a `guint64` rather than `gint64`?

Reply:
It can't be outside the positive range of a gint64 because `times_us` is a gint64 because GLib represents such timestamps as a signed gint64. I don't think this looks any better:

```c
  g_return_if_fail (delta_us < G_MAXINT64);
  g_return_if_fail (self->times_us[MOCK_CLOCK_TYPE_MONOTONIC] <= G_MAXINT64 - (gint64) delta_us);
```

(The cast is needed because `times_us[]` is signed; otherwise the compiler gets sad.)

Reply:
Could `times_us` also be `guint64`?

Reply:
It could, but there is also code in `emer-daemon` which retrieves the current monotonic time, which is a `gint64`. Currently it gets the real time, but I would like later to mock that, too. I think it makes sense to be consistent with the way the regular GLib functions represent time.

Reply:
> I think it makes sense to be consistent with the way the regular GLib functions represent time.

Sure, that’s a valid argument and I’m not going to push further. Though I will point out that, as far as I am aware, it’s a historical accident that the GLib functions use `gint64` rather than `guint64`.